indexer/SymbolFormatter.cc

```diff
@@ -3,19 +3,123 @@
 
 #include "absl/container/flat_hash_map.h"
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
 #include "absl/strings/str_replace.h"
 
 #include "clang/AST/Decl.h"
 #include "clang/Basic/SourceLocation.h"
 #include "clang/Basic/SourceManager.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "scip/scip.pb.h"
 
 #include "indexer/Enforce.h"
 #include "indexer/Hash.h"
 #include "indexer/LlvmAdapter.h"
 #include "indexer/SymbolFormatter.h"
 
+namespace {
+struct Escaped {
+  std::string_view text;
+};
+} // namespace
+
+namespace llvm {
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, Escaped s) {
+  bool needsEscape = false;
+  bool needsBacktickEscape = false;
+  for (auto &c : s.text) {
+    if (('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z')
+        || ('0' <= c && c <= '9') || c == '+' || c == '-' || c == '$'
+        || c == '_') {
+      continue;
+    }
+    needsBacktickEscape = c == '`';
+    needsEscape = true;
+  }
+  if (needsEscape) {
+    if (needsBacktickEscape) {
+      return os << '`' << absl::StrReplaceAll(s.text, {{"`", "``"}}) << '`';
+    }
+    return os << '`' << s.text << '`';
+  }
+  return os << s.text;
+}
+} // namespace llvm
+
+static void addSpaceEscaped(llvm::raw_ostream &out, std::string_view in) {
+  if (absl::StrContains(in, ' ')) {
+    out << absl::StrReplaceAll(in, {{" ", "  "}});
+    return;
+  }
+  out << in;
+}
+
 namespace scip_clang {
 
+void DescriptorBuilder::formatTo(llvm::raw_ostream &out) const {
+  // See https://github.com/sourcegraph/scip/blob/main/scip.proto#L104-L125
+  switch (this->suffix) {
+  case scip::Descriptor::Namespace:
+    out << ::Escaped{this->name} << '/';
+    break;
+  case scip::Descriptor::Type:
+    out << ::Escaped{this->name} << '#';
+    break;
+  case scip::Descriptor::Term:
+    out << ::Escaped{this->name} << '.';
+    break;
+  case scip::Descriptor::Meta:
+    out << ::Escaped{this->name} << ':';
+    break;
+  case scip::Descriptor::Method:
+    out << ::Escaped{this->name} << '(' << ::Escaped{this->disambiguator}
+        << ").";
+    break;
+  case scip::Descriptor::TypeParameter:
+    out << '[' << ::Escaped{this->name} << ']';
+    break;
+  case scip::Descriptor::Parameter:
+    out << '(' << ::Escaped{this->name} << ')';
+    break;
+  case scip::Descriptor::Macro:
+    out << ::Escaped{this->name} << '!';
+    break;
+  default:
+    ENFORCE(false, "unknown descriptor suffix %s",
+            scip::Descriptor::Suffix_Name(this->suffix));
+  }
+}
+
+void SymbolBuilder::formatTo(std::string &buf) const {
+  llvm::raw_string_ostream out(buf);
+  out << "cxx . "; // scheme manager
+  for (auto text : {this->packageName, this->packageVersion}) {
+    if (text.empty()) {
+      out << ". ";
+      continue;
+    }
+    ::addSpaceEscaped(out, text);
+    out << ' ';
+  }
+  for (auto &descriptor : this->descriptors) {
+    descriptor.formatTo(out);
+  }
+}
+
+std::string
+SymbolBuilder::formatContextual(std::string_view contextSymbol,
```

Comment:
`contextSymbol` should be a const reference.

Reply:
Do you mean using `const std::string &contextSymbol` instead of `std::string_view`? That would not work well with the rest of the code because other methods return `optional<string_view>` or `string_view` values.